File: scala/src/main/scala/org/dhallj/s/visitor/ConstantVisitor.scala
package org.dhallj.s.visitor

import java.net.URI
import java.nio.file.Path
import org.dhallj.core.{Expr, Operator, Source}
import org.dhallj.s.Visitor

trait ConstantVisitor[A] extends Visitor[A] {
  def constantValue: A
  def onDouble(value: Double): A = constantValue
  def onNatural(value: BigInt): A = constantValue
  def onInteger(value: BigInt): A = constantValue
  def onBuiltIn(namee: String): A = constantValue
  def onIdentifier(name: String, index: Option[Long]): A = constantValue
  def onText(parts: Iterable[String], interpolated: Iterable[Expr]): A = constantValue
  def onApplication(base: Expr, arg: Expr): A = constantValue
  def onOperatorApplication(operator: Operator, lhs: Expr, rhs: Expr): A = constantValue
  def onIf(cond: Expr, thenValue: Expr, elseValue: Expr): A = constantValue
  def onLambda(param: String, input: Expr, result: Expr): A = constantValue
  def onPi(param: Option[String], input: Expr, result: Expr): A = constantValue
  def onAssert(base: Expr): A = constantValue
  def onFieldAccess(base: Expr, fieldName: String): A = constantValue
  def onProjection(base: Expr, fieldNames: Iterable[String]): A = constantValue
  def onProjectionByType(base: Expr, tpe: Expr): A = constantValue
  def onRecord(fields: Iterable[(String, Expr)], size: Int): A = constantValue
  def onRecordType(fields: Iterable[(String, Expr)], size: Int): A = constantValue
  def onUnionType(fields: Iterable[(String, Option[Expr])], size: Int): A = constantValue
  def onNonEmptyList(values: Iterable[Expr], size: Int): A = constantValue
  def onEmptyList(tpe: Expr): A = constantValue
  def onLet(name: String, tpe: Option[Expr], value: Expr, body: Expr): A = constantValue
  def onAnnotated(base: Expr, tpe: Expr): A = constantValue
  def onMerge(handlers: Expr, union: Expr, tpe: Option[Expr]): A = constantValue
  def onToMap(base: Expr, tpe: Option[Expr]): A = constantValue
  def onLocalImport(path: Path, mode: Expr.ImportMode, hash: Option[Array[Byte]]): A = constantValue
  def onEnvImport(value: String, mode: Expr.ImportMode, hash: Option[Array[Byte]]): A = constantValue
  def onRemoteImport(url: URI, using: Option[Expr], mode: Expr.ImportMode, hash: Option[Array[Byte]]): A = constantValue
  def onMissingImport(mode: Expr.ImportMode, hash: Option[Array[Byte]]): A = constantValue
}

object ConstantVisitor {
  class Optional[A] extends ConstantVisitor[Option[A]] {
    def constantValue: Option[A] = None
    def onNote(base: Expr, source: Source): Option[A] = base.accept(this)
  }
}

File: core/src/main/java/org/dhallj/core/visitor/ConstantVisitor.java
package org.dhallj.core.visitor;

import java.math.BigInteger;
import java.net.URI;
import java.nio.file.Path;
import java.util.Map.Entry;
import org.dhallj.core.Expr;
import org.dhallj.core.Import;
import org.dhallj.core.Operator;
import org.dhallj.core.Source;
import org.dhallj.core.Visitor;

/**
 * Represents a function from a Dhall expression to a value that always returns the same value.
 *
 * <p>This is a convenience class designed to help with implementations that have a default value
 * for most cases.
 *
 * <p>Note that both the internal and external implementations recurse through note layers.
 *
 * @param I The internal result type used while recursing
 * @param A The final result type
 */
public class ConstantVisitor<I, A> implements Visitor<I, A> {
  private final A returnValue;

  protected A getReturnValue() {
    return this.returnValue;
  }

  ConstantVisitor(A value) {
    this.returnValue = value;
  }

  public static class External<A> extends ConstantVisitor<Expr, A> implements ExternalVisitor<A> {

    public External(A value) {
      super(value);
    }

    @Override
    public A onNote(Expr base, Source source) {
      return base.acceptExternal(this);
    }
  }

  public A onDoubleLiteral(double value) {
    return this.getReturnValue();
  }

  public A onNaturalLiteral(BigInteger value) {
    return this.getReturnValue();
  }

  public A onIntegerLiteral(BigInteger value) {
    return this.getReturnValue();
  }

  public A onTextLiteral(String[] parts, Iterable<I> interpolated) {
    return this.getReturnValue();
  }

  public A onApplication(I base, I arg) {
    return this.getReturnValue();
  }

  public A onOperatorApplication(Operator operator, I lhs, I rhs) {
    return this.getReturnValue();
  }

  public A onIf(I cond, I thenValue, I elseValue) {
    return this.getReturnValue();
  }

  public A onLambda(String param, I input, I result) {
    return this.getReturnValue();
  }

  public A onPi(String param, I input, I result) {
    return this.getReturnValue();
  }

  public A onAssert(I base) {
    return this.getReturnValue();
  }

  public A onFieldAccess(I base, String fieldName) {
    return this.getReturnValue();
  }

  public A onProjection(I base, String[] fieldNames) {
    return this.getReturnValue();
  }

  public A onProjectionByType(I base, I tpe) {
    return this.getReturnValue();
  }

  public A onBuiltIn(String name) {
    return this.getReturnValue();
  }

  public A onIdentifier(String value, long index) {
    return this.getReturnValue();
  }

  public A onRecordLiteral(Iterable<Entry<String, I>> fields, int size) {
    return this.getReturnValue();
  }

  public A onRecordType(Iterable<Entry<String, I>> fields, int size) {
    return this.getReturnValue();
  }

  public A onUnionType(Iterable<Entry<String, I>> fields, int size) {
    return this.getReturnValue();
  }

  public A onNonEmptyListLiteral(Iterable<I> values, int size) {
    return this.getReturnValue();
  }

  public A onEmptyListLiteral(I tpe) {
    return this.getReturnValue();
  }

  public A onLet(String name, I type, I value, I body) {
    return this.getReturnValue();
  }

  public A onAnnotated(I base, I tpe) {
    return this.getReturnValue();
  }

  public A onToMap(I base, I tpe) {
    return this.getReturnValue();
  }

  public A onMerge(I left, I right, I tpe) {

    return this.getReturnValue();
  }

  public A onLocalImport(Path path, Import.Mode mode, byte[] hash) {
    return this.getReturnValue();
  }

  public A onRemoteImport(URI url, I using, Import.Mode mode, byte[] hash) {
    return this.getReturnValue();
  }

  public A onEnvImport(String value, Import.Mode mode, byte[] hash) {
    return this.getReturnValue();
  }

  public A onMissingImport(Import.Mode mode, byte[] hash) {
    return this.getReturnValue();
  }

  public A onNote(I base, Source source) {
    return this.getReturnValue();
  }
}
